plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -852,6 +854,35 @@ entryExitTracing lamName displayName e ty =
    an unfolding.
 -}
 
+takeHaskellList
+  :: forall uni fun m ann
+   . (CompilingDefault uni fun m ann)
+  => GHC.CoreExpr
+  -> m [PIRTerm uni fun]
+takeHaskellList = withOrWithoutTick buildList
+  where
+    err =
+      throwPlain $
+        CompilationError "Unexpected expression is given where to be Haskell list constructor is expected"
+
+    withOrWithoutTick :: (GHC.CoreExpr -> a) -> GHC.CoreExpr -> a
+    withOrWithoutTick f (GHC.Tick _ e) = f e
+    withOrWithoutTick f e              = f e
+
+    buildList (GHC.App (GHC.App (GHC.App (GHC.Var _con) _ty) e) _) = traverse compileExpr [e]
+    buildList (GHC.App (GHC.App _build _ty) (GHC.Lam _tyArg (GHC.Lam con (GHC.Lam nil li)))) =
```

Comment:
Add some comments, showing examples of what the CoreExpr looks like in different cases (empty list, singleton, multiple elements).